src/hotspot/share/jfr/leakprofiler/chains/pathToGcRootsOperation.cpp

```diff
@@ -129,3 +136,16 @@ void PathToGcRootsOperation::doit() {
   EventEmitter emitter(GranularTimer::start_time(), GranularTimer::end_time());
   emitter.write_events(_sampler, _edge_store, _emit_all);
 }
+
+bool PathToGcRootsOperation::is_safe() {
+#if INCLUDE_SHENANDOAHGC
+  if (UseShenandoahGC) {
+    // This operation uses mark words to track objects. While the operation
+    // would restore the mark words after completion, it would interact with
+    // mark word uses by Shenandoah itself, if we hit the op during the concurrent
+    // GC cycle.
+    return ShenandoahHeap::heap()->is_idle();
```

Comment:
I think you need to include shenandoahHeap.inline.hpp for this? Also, do we need to block this during all of GC, even marking? Would it make sense to only block during evac/updaterefs?

Reply:
One of my attempts _did_ have a more sharp `!has_forwarded_objects()` check. But then I chickened out from specializing this, and instead trusting the more catch-all `is_idle()`. We can do a more precise test and relax it if we see more problems.

And right, we need a sharper include here, added.

See new commit.